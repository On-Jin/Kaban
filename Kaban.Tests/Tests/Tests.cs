using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using Kaban.Data;
using Kaban.Models;
using Kaban.Tests.Setup;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit.Abstractions;

namespace Kaban.Tests.Tests;

[Collection(nameof(WebAppCollectionFixture))]
public class Tests : IAsyncLifetime
{
    private readonly ITestOutputHelper _testOutputHelper;
    private readonly HttpClient _httpClientDiscord;
    private readonly HttpClient _httpClientShadow;

    private readonly Func<Task> _resetDatabase;
    private readonly WebAppFactory _factory;

    public Tests(WebAppFactory factory, ITestOutputHelper testOutputHelper)
    {
        _factory = factory;
        _testOutputHelper = testOutputHelper;
        _resetDatabase = factory.ResetDatabaseAsync;

        _httpClientDiscord = factory.CreateClient();
        _httpClientDiscord.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue(TestAuthHandler.AuthenticationScheme);
        _httpClientShadow = factory.CreateClient();

        _httpClientShadow.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue(TestAuthHandler.AuthenticationScheme);
        _httpClientShadow.DefaultRequestHeaders.Add(TestHelper.HeaderUserGuid, TestHelper.GuidShadowAuth);
    }

    public Task InitializeAsync() => Task.CompletedTask;

    public Task DisposeAsync() => _resetDatabase();

    private async Task GenerateKabanBoardToShadowUser(AppDbContext db)
    {
        var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
        var boards = await GenerateHelper.GenerateDefaultKabanBoards();
        user.Boards.AddRange(boards);
        await db.SaveChangesAsync();
    }

    [Fact]
    public async Task DeleteOneBoardTestCascade()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        await GenerateKabanBoardToShadowUser(db);

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            user.Boards.RemoveAt(0);
            await db.SaveChangesAsync();
        }
        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            JsonSerializer.Serialize(user, TestHelper.JsonSerializerOptions).MatchSnapshot();
            Assert.Equal(2, db.Boards.Count());
        }
    }

    [Fact]
    public async Task AddGeneratedKabanBoardsTest()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        await GenerateKabanBoardToShadowUser(db);

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            JsonSerializer.Serialize(user, TestHelper.JsonSerializerOptions).MatchSnapshot();
        }
    }

    [Fact]
    public async Task Test1()
    {
        await _resetDatabase();

        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            var boards = await GenerateHelper.GenerateDefaultKabanBoards();
            user.Boards.AddRange(boards);
            await db.SaveChangesAsync();
        }

        {
            _testOutputHelper.WriteLine(db.Boards.ToList().Count.ToString());
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            user.Boards.RemoveAt(0);
            await db.SaveChangesAsync();
            _testOutputHelper.WriteLine(db.Boards.ToList().Count.ToString());
        }

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            _testOutputHelper.WriteLine(JsonSerializer.Serialize(user, TestHelper.JsonSerializerOptions));
        }
        {
            _testOutputHelper.WriteLine(db.Users.ToList().Count.ToString());
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            db.Users.Remove(user);
            await db.SaveChangesAsync();
            _testOutputHelper.WriteLine(db.Users.ToList().Count.ToString());
        }
    }
}