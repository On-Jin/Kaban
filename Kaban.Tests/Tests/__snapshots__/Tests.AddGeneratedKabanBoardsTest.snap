{
  "Id": "158ef677-f7a1-424a-bea4-0d6d8ec95717",
  "DiscordId": null,
  "DiscordAvatar": null,
  "DiscordUsername": null,
  "Boards": [
    {
      "Id": 1,
      "Name": "Platform Launch",
      "Columns": [
        {
          "Id": 1,
          "Name": "Todo",
          "MainTasks": [
            {
              "Id": 1,
              "Title": "Build UI for onboarding flow",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 1,
                  "Title": "Sign up page",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 1
                },
                {
                  "Id": 2,
                  "Title": "Sign in page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 1
                },
                {
                  "Id": 3,
                  "Title": "Welcome page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 1
                }
              ],
              "Column": null,
              "ColumnId": 1
            },
            {
              "Id": 2,
              "Title": "Build UI for search",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 4,
                  "Title": "Search page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 2
                }
              ],
              "Column": null,
              "ColumnId": 1
            },
            {
              "Id": 3,
              "Title": "Build settings UI",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 5,
                  "Title": "Account page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 3
                },
                {
                  "Id": 6,
                  "Title": "Billing page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 3
                }
              ],
              "Column": null,
              "ColumnId": 1
            },
            {
              "Id": 4,
              "Title": "QA and test all major user journeys",
              "Description": "Once we feel version one is ready, we need to rigorously test it both internally and externally to identify any major gaps.",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 7,
                  "Title": "Internal testing",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 4
                },
                {
                  "Id": 8,
                  "Title": "External testing",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 4
                }
              ],
              "Column": null,
              "ColumnId": 1
            }
          ],
          "Board": null,
          "BoardId": 1
        },
        {
          "Id": 2,
          "Name": "Doing",
          "MainTasks": [
            {
              "Id": 5,
              "Title": "Design settings and search pages",
              "Description": "",
              "Status": 1,
              "SubTasks": [
                {
                  "Id": 9,
                  "Title": "Settings - Account page",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 5
                },
                {
                  "Id": 10,
                  "Title": "Settings - Billing page",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 5
                },
                {
                  "Id": 11,
                  "Title": "Search page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 5
                }
              ],
              "Column": null,
              "ColumnId": 2
            },
            {
              "Id": 6,
              "Title": "Add account management endpoints",
              "Description": "",
              "Status": 1,
              "SubTasks": [
                {
                  "Id": 12,
                  "Title": "Upgrade plan",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 6
                },
                {
                  "Id": 13,
                  "Title": "Cancel plan",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 6
                },
                {
                  "Id": 14,
                  "Title": "Update payment method",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 6
                }
              ],
              "Column": null,
              "ColumnId": 2
            },
            {
              "Id": 7,
              "Title": "Design onboarding flow",
              "Description": "",
              "Status": 1,
              "SubTasks": [
                {
                  "Id": 15,
                  "Title": "Sign up page",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 7
                },
                {
                  "Id": 16,
                  "Title": "Sign in page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 7
                },
                {
                  "Id": 17,
                  "Title": "Welcome page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 7
                }
              ],
              "Column": null,
              "ColumnId": 2
            },
            {
              "Id": 8,
              "Title": "Add search enpoints",
              "Description": "",
              "Status": 1,
              "SubTasks": [
                {
                  "Id": 18,
                  "Title": "Add search endpoint",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 8
                },
                {
                  "Id": 19,
                  "Title": "Define search filters",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 8
                }
              ],
              "Column": null,
              "ColumnId": 2
            },
            {
              "Id": 9,
              "Title": "Add authentication endpoints",
              "Description": "",
              "Status": 1,
              "SubTasks": [
                {
                  "Id": 20,
                  "Title": "Define user model",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 9
                },
                {
                  "Id": 21,
                  "Title": "Add auth endpoints",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 9
                }
              ],
              "Column": null,
              "ColumnId": 2
            },
            {
              "Id": 10,
              "Title": "Research pricing points of various competitors and trial different business models",
              "Description": "We know what we\u0027re planning to build for version one. Now we need to finalise the first pricing model we\u0027ll use. Keep iterating the subtasks until we have a coherent proposition.",
              "Status": 1,
              "SubTasks": [
                {
                  "Id": 22,
                  "Title": "Research competitor pricing and business models",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 10
                },
                {
                  "Id": 23,
                  "Title": "Outline a business model that works for our solution",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 10
                },
                {
                  "Id": 24,
                  "Title": "Talk to potential customers about our proposed solution and ask for fair price expectancy",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 10
                }
              ],
              "Column": null,
              "ColumnId": 2
            }
          ],
          "Board": null,
          "BoardId": 1
        },
        {
          "Id": 3,
          "Name": "Done",
          "MainTasks": [
            {
              "Id": 11,
              "Title": "Conduct 5 wireframe tests",
              "Description": "Ensure the layout continues to make sense and we have strong buy-in from potential users.",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 25,
                  "Title": "Complete 5 wireframe prototype tests",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 11
                }
              ],
              "Column": null,
              "ColumnId": 3
            },
            {
              "Id": 12,
              "Title": "Create wireframe prototype",
              "Description": "Create a greyscale clickable wireframe prototype to test our asssumptions so far.",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 26,
                  "Title": "Create clickable wireframe prototype in Balsamiq",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 12
                }
              ],
              "Column": null,
              "ColumnId": 3
            },
            {
              "Id": 13,
              "Title": "Review results of usability tests and iterate",
              "Description": "Keep iterating through the subtasks until we\u0027re clear on the core concepts for the app.",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 27,
                  "Title": "Meet to review notes from previous tests and plan changes",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 13
                },
                {
                  "Id": 28,
                  "Title": "Make changes to paper prototypes",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 13
                },
                {
                  "Id": 29,
                  "Title": "Conduct 5 usability tests",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 13
                }
              ],
              "Column": null,
              "ColumnId": 3
            },
            {
              "Id": 14,
              "Title": "Create paper prototypes and conduct 10 usability tests with potential customers",
              "Description": "",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 30,
                  "Title": "Create paper prototypes for version one",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 14
                },
                {
                  "Id": 31,
                  "Title": "Complete 10 usability tests",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 14
                }
              ],
              "Column": null,
              "ColumnId": 3
            },
            {
              "Id": 15,
              "Title": "Market discovery",
              "Description": "We need to define and refine our core product. Interviews will help us learn common pain points and help us define the strongest MVP.",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 32,
                  "Title": "Interview 10 prospective customers",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 15
                }
              ],
              "Column": null,
              "ColumnId": 3
            },
            {
              "Id": 16,
              "Title": "Competitor analysis",
              "Description": "",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 33,
                  "Title": "Find direct and indirect competitors",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 16
                },
                {
                  "Id": 34,
                  "Title": "SWOT analysis for each competitor",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 16
                }
              ],
              "Column": null,
              "ColumnId": 3
            },
            {
              "Id": 17,
              "Title": "Research the market",
              "Description": "We need to get a solid overview of the market to ensure we have up-to-date estimates of market size and demand.",
              "Status": 2,
              "SubTasks": [
                {
                  "Id": 35,
                  "Title": "Write up research analysis",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 17
                },
                {
                  "Id": 36,
                  "Title": "Calculate TAM",
                  "IsCompleted": true,
                  "MainTask": null,
                  "MainTaskId": 17
                }
              ],
              "Column": null,
              "ColumnId": 3
            }
          ],
          "Board": null,
          "BoardId": 1
        }
      ]
    },
    {
      "Id": 2,
      "Name": "Marketing Plan",
      "Columns": [
        {
          "Id": 4,
          "Name": "Todo",
          "MainTasks": [
            {
              "Id": 18,
              "Title": "Plan Product Hunt launch",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 37,
                  "Title": "Find hunter",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 38,
                  "Title": "Gather assets",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 39,
                  "Title": "Draft product page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 40,
                  "Title": "Notify customers",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 41,
                  "Title": "Notify network",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 42,
                  "Title": "Launch!",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                }
              ],
              "Column": null,
              "ColumnId": 4
            },
            {
              "Id": 19,
              "Title": "Share on Show HN",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 43,
                  "Title": "Draft out HN post",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 19
                },
                {
                  "Id": 44,
                  "Title": "Get feedback and refine",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 19
                },
                {
                  "Id": 45,
                  "Title": "Publish post",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 19
                }
              ],
              "Column": null,
              "ColumnId": 4
            },
            {
              "Id": 20,
              "Title": "Write launch article to publish on multiple channels",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 46,
                  "Title": "Write article",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                },
                {
                  "Id": 47,
                  "Title": "Publish on LinkedIn",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                },
                {
                  "Id": 48,
                  "Title": "Publish on Inndie Hackers",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                },
                {
                  "Id": 49,
                  "Title": "Publish on Medium",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                }
              ],
              "Column": null,
              "ColumnId": 4
            }
          ],
          "Board": null,
          "BoardId": 2
        },
        {
          "Id": 5,
          "Name": "Doing",
          "MainTasks": [],
          "Board": null,
          "BoardId": 2
        },
        {
          "Id": 6,
          "Name": "Done",
          "MainTasks": [],
          "Board": null,
          "BoardId": 2
        }
      ]
    },
    {
      "Id": 3,
      "Name": "Roadmap",
      "Columns": [
        {
          "Id": 7,
          "Name": "Now",
          "MainTasks": [
            {
              "Id": 21,
              "Title": "Launch version one",
              "Description": "",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 50,
                  "Title": "Launch privately to our waitlist",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 21
                },
                {
                  "Id": 51,
                  "Title": "Launch publicly on PH, HN, etc.",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 21
                }
              ],
              "Column": null,
              "ColumnId": 7
            },
            {
              "Id": 22,
              "Title": "Review early feedback and plan next steps for roadmap",
              "Description": "Beyond the initial launch, we\u0027re keeping the initial roadmap completely empty. This meeting will help us plan out our next steps based on actual customer feedback.",
              "Status": 0,
              "SubTasks": [
                {
                  "Id": 52,
                  "Title": "Interview 10 customers",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 22
                },
                {
                  "Id": 53,
                  "Title": "Review common customer pain points and suggestions",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 22
                },
                {
                  "Id": 54,
                  "Title": "Outline next steps for our roadmap",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 22
                }
              ],
              "Column": null,
              "ColumnId": 7
            }
          ],
          "Board": null,
          "BoardId": 3
        },
        {
          "Id": 8,
          "Name": "Next",
          "MainTasks": [],
          "Board": null,
          "BoardId": 3
        },
        {
          "Id": 9,
          "Name": "Later",
          "MainTasks": [],
          "Board": null,
          "BoardId": 3
        }
      ]
    }
  ]
}
